### PR TITLE
[backend] Integration of TournesolBot into the backend

### PR DESCRIPTION
**Pull request by aidanjungo**
More details in #512

### CODE REVIEW

File: backend/twitterbot/uploader_twitter_account.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        return None
+
+    uploader_url = f"https://www.youtube.com/channel/{uploader_id}"
+
+    r = requests.get(uploader_url, headers={"user-agent": "curl/7.68.0"})
+
+    twitter_names = re.findall(r"twitter.com\%2F(.*?)\"", r.text)
+    twitter_names = [name for name in twitter_names if name != ""]
+
+    if len(set([name.lower() for name in twitter_names])) == 1:
```
Comment (amatissart) [inline backend/twitterbot/uploader_twitter_account.py:42]: The list is not necessary, `set()` accepts an iterator:
```suggestion
    if len(set(name.lower() for name in twitter_names)) == 1:
```

File: backend/twitterbot/uploader_twitter_account.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    url_video = f"https://www.youtube.com/watch?v={video_id}"
+
+    r = requests.get(url_video)
+
+    if r.status_code != 200:
+        raise ConnectionError(f"Error getting video id: {r.status_code}")
+
+    try:
+        uploader_id = r.text.split("/channel/")[1].split('"')[0]
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/uploader_twitter_account.py:15]: I will try to get the uploader_id from the YouTube api.

**Comment by aidanjungo** [comment]
@amatissart if I did it correctly, it seems better like that to get the uploader_id

File: backend/twitterbot/tournesolbot.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+            resp = twitterbot.api.update_status(tweet_text)
+
+            # Add the video to the TweetInfo table
+            TweetInfo.objects.create(
+                video=video,
+                tweet_id=resp.id,
+                datetime_tweet=twitterbot.api.now(),
+                bot_name=bot_name,
+            )
```
Comment (Vikka) [inline backend/twitterbot/tournesolbot.py:152]: PEP20, flat is better than nested.

```suggestion
    if not debug:
        return

    print("Today's video to tweet will be:")
    print(tweet_text)
    
    confirmation = input("\nWould you like to tweet that? (y/n)")
    if confirmation != "y":
        return

    resp = twitterbot.api.update_status(tweet_text)
    
    # Add the video to the TweetInfo table
    TweetInfo.objects.create(
        video=video,
        tweet_id=resp.id,
        datetime_tweet=twitterbot.api.now(),
        bot_name=bot_name,
    )
```

File: backend/twitterbot/uploader_twitter_account.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    """Get Twitter account from uploader id"""
+
+    uploader_url = f"https://www.youtube.com/channel/{channel_id}"
+
+    resp = requests.get(uploader_url, headers={"user-agent": "curl/7.68.0"})
+
+    twitter_names = re.findall(r"twitter.com\%2F(.*?)\"", resp.text)
+    twitter_names = [name for name in twitter_names if name != ""]
+
+    if len(set(name.lower() for name in twitter_names)) == 1:
```
Comment (Vikka) [inline backend/twitterbot/uploader_twitter_account.py:21]: set comprehension is prefered as generator expresion casted as set
```suggestion
    if len({name.lower() for name in twitter_names}) == 1:
```

**Review by Vikka — CHANGES_REQUESTED** [review]
Hi,
Mostly layout or pythonicness of the code.

File: backend/twitterbot/tests/test_get_twitter_account.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+from unittest.mock import MagicMock, patch
+
+
+@patch("twitterbot.uploader_twitter_account.requests.get")
+def test_get_twitter_account_from_channel_id(mock_requests):
+    """Test function for get_twitter_account_from_channel_id"""
+
+    # Correct response from youtube
+    with open("./twitterbot/tests/mock_resp.txt") as f:
+        mock_resp = f.read()
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/tests/test_get_twitter_account.py:11]: 👍

File: backend/twitterbot/tournesolbot.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+from twitterbot.twitter_api import TwitterBot
+from twitterbot.uploader_twitter_account import get_twitter_account_from_video_id
+
+
+def get_best_criteria(video, n):
+    """Get the n best criteria"""
+
+    criteria_list = video.criteria_scores.filter(poll__name="videos").order_by(
+        "-score"
+    )[:n]
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/tournesolbot.py:17]: I use Black so I let it like that for now.

File: backend/twitterbot/tournesolbot.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    if not twitter_account:
+        twitter_account = f"'{uploader}'"
+
+    # Get two best criteria and criteria dict name
+    crit1, crit2 = get_best_criteria(video, 2)
+    crit_dict = dict(
+        CriteriaLocale.objects.filter(language=language).values_list(
+            "criteria__name", "label"
+        )
+    )
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/tournesolbot.py:41]: I use Black so I let it like that for now.

File: backend/twitterbot/tournesolbot.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+def get_video_recommendations(language):
+    """Find a list of good video to recommend with some quality criteria."""
+
+    # Filter videos with some quality criteria
+    tweetable_videos = Entity.objects.filter(
+        add_time__lte=time_ago(days=settings.DAYS_TOO_RECENT),
+        metadata__publication_date__gte=time_ago(
+            days=settings.DAYS_TOO_OLD
+        ).isoformat(),
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/tournesolbot.py:84]: I use Black so I let it like that for now.

File: backend/twitterbot/tournesolbot.py (hunk 3 above)
Comment (aidanjungo) [reply backend/twitterbot/tournesolbot.py]: Thanks, it is indeed better as you propose :)

File: backend/twitterbot/uploader_twitter_account.py (hunk 4 above)
Comment (aidanjungo) [reply backend/twitterbot/uploader_twitter_account.py]: Thanks

**Comment by Vikka** [comment]
@aidanjungo don't hesitate to accept my suggestions with github as it will mention me in co-author of the commit involving my suggestions.

File: infra/ansible/roles/django/templates/settings.yaml.j2 [hunk 9]
```diff
@@ -54,3 +54,19 @@ YOUTUBE_API_KEY: {{youtube_api_key}}
...
+        "ACCESS_TOKEN": "{{access_token_twitterbot_fr}}",
+        "ACCESS_TOKEN_SECRET": "{{access_token_secret_twitterbot_fr}}",
+    }
+    "@TournesolBot": {
+        "LANGUAGE": "en",
+        "CONSUMER_KEY": "{{consumer_key_twitterbot_en}}",
+        "CONSUMER_SECRET": "{{consumer_secret_twitterbot_en}}",
+        "ACCESS_TOKEN": "{{access_token_twitterbot_en}}",
+        "ACCESS_TOKEN_SECRET": "{{access_token_secret_twitterbot_en}}",
+    }
```
Comment (amatissart) [inline infra/ansible/roles/django/templates/settings.yaml.j2:72]: I think that Ansible will fail when using this template including new variables that are not defined in the inventory yet. 

I would recommend to move this change to #702, to be able to merge and deploy these 2 PRs step by step.

**Review by amatissart — COMMENTED** [review]
I have a few questions, and there are some minor issues to address before merging. But overall it looks good :+1:

File: backend/twitterbot/tournesolbot.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    tweetable_videos = get_video_recommendations(language=twitterbot.language)
+    if not tweetable_videos:
+        print("No video reach the criteria to be tweeted today!!!")
+        return
+
+    video = select_a_video(tweetable_videos)
+    tweet_text = prepare_tweet(video)
+
+    if not debug:
+        return
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/tournesolbot.py:136]: Yes it was just to be sure during the first tests and only use it in debug until we are sure everything works.

File: backend/twitterbot/tournesolbot.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    """Find a list of good video to recommend with some quality criteria."""
+
+    # Filter videos with some quality criteria
+    tweetable_videos = Entity.objects.filter(
+        add_time__lte=time_ago(days=settings.DAYS_TOO_RECENT),
+        metadata__publication_date__gte=time_ago(
+            days=settings.DAYS_TOO_OLD
+        ).isoformat(),
+        rating_n_contributors__gt=settings.MIN_NB_CONTRIBUTORS,
+        rating_n_ratings__gt=settings.MIN_NB_RATINGS,
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/tournesolbot.py]: done

File: backend/twitterbot/tournesolbot.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        tournesol_score__gte=settings.MIN_TOURNESOL_SCORE,
+        criteria_scores__criteria="reliability",
+        criteria_scores__score__gt=settings.MIN_RELIABILITY_SCORE,
+        tweets=None,
+    )
+
+    # Exclude video recently tweeted by the same uploader
+    exclude_uploader = TweetInfo.objects.filter(
+        datetime_tweet__gte=time_ago(days=settings.DAYS_ALREADY_TWEETED_UPLOADER)
+    ).values_list("video__metadata__uploader", flat=True)
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/tournesolbot.py]: done

File: backend/twitterbot/tournesolbot.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    video_id = video.metadata["video_id"]
+
+    # Get twitter account
+    twitter_account = get_twitter_account_from_video_id(video_id)
+
+    if not twitter_account:
+        twitter_account = f"'{uploader}'"
+
+    # Get two best criteria and criteria dict name
+    crit1, crit2 = get_best_criteria(video, 2)
```
[the remark this answers is not in the document]
Comment (aidanjungo) [reply backend/twitterbot/tournesolbot.py:39]: Now get_best_criteria raises an error when there is not enough criteria (should not happen too often thanks to the other filters)

**Review by amatissart — APPROVED** [review]

**Pull request merged**